File: src/customtypes/type_indentationtracker.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# IMPORTS
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

from __future__ import annotations;

from src.core.utils import sizeOfIndent;

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# GLOBAL VARIABLES
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#

# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# CLASS indentation tracker
# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

class IndentationTracker(object):
    pattern:    str;
    symb:       str;
    _reference: int;
    level:     int;

    def __init__(
        self,
        symb: str,
        pattern: str
    ):
        self.symb = symb;
        self.pattern = pattern;
        self.initOffset();
        return;

    def relativeOffset(self, s: str):
        return max(sizeOfIndent(s, indentsymb=self.symb) - self._reference, 0);

    def initOffset(self, s: str = '') -> int:
        self.level      = 0;
        self._reference = sizeOfIndent(s, indentsymb=self.symb);
        return self._reference;

    def setOffset(self, s: str) -> int:
        n = self.relativeOffset(s);
        self.level = n;
        return self.level;

    def decrOffset(self) -> int:
        self.level = max(self.level - 1, 0);
        return self.level;

    def incrOffset(self) -> int:
        self.level += 1;
        return self.level;

File: golang/internal/types/type_indentationtracker.go
package types

/* ---------------------------------------------------------------- *
 * IMPORTS
 * ---------------------------------------------------------------- */

import (
	"phpytex/internal/core/utils"
)

/* ---------------------------------------------------------------- *
 * TYPES
 * ---------------------------------------------------------------- */

type IndentationTracker struct {
	Symb      string
	Pattern   string
	reference int
	level     int
}

/* ---------------------------------------------------------------- *
 * METHODS
 * ---------------------------------------------------------------- */

func NewIndentationTracker(Symb string, Pattern string) IndentationTracker {
	var it IndentationTracker = IndentationTracker{Symb: Symb, Pattern: Pattern}
	it.Init()
	return it
}

func (it *IndentationTracker) Init(options ...string) {
	var reference string = utils.GetArrayStringValue(&options, 0, "")
	it.level = 0
	it.reference = utils.SizeOfIndent(reference, it.Symb)
}

func (it IndentationTracker) GetLevel() int {
	return it.level
}

func (it *IndentationTracker) relativeOffset(s string) int {
	var n int = utils.SizeOfIndent(s, it.Symb) - it.reference
	if n < 0 {
		n = 0
	}
	return n
}

func (it *IndentationTracker) SetOffset(s string) {
	it.level = it.relativeOffset(s)
}

func (it *IndentationTracker) DecrOffset() {
	it.level--
	if it.level < 0 {
		it.level = 0
	}
}

func (it *IndentationTracker) Incroffset() {
	it.level++
}
